pkg/operator/controller/ingress/status.go

```diff
@@ -556,6 +574,50 @@ func computeIngressUpgradeableCondition(ic *operatorv1.IngressController, deploy
 	}
 }
 
+// checkDefaultCertificate returns an error value indicating whether the default
+// certificate is safe for upgrades.  In particular, if the current default
+// certificate specifies a Subject Alternative Name (SAN) for the ingress
+// domain, then it is safe to upgrade, and the return value is nil.  Otherwise,
+// if the certificate has a legacy Common Name (CN) for the ingress domain and
+// no SAN for the same, then the return value is an error indicating that the
+// certificate must be replaced by one with a SAN before upgrading is allowed.
+// This check is necessary because OpenShift 4.10 is built using Go 1.17, which
+// rejects certificates without SANs.  Note that this function only checks the
+// validity of the certificate insofar as it affects upgrades.  See
+// <https://bugzilla.redhat.com/show_bug.cgi?id=2057762>.  This check can be
+// removed after OpenShift 4.9.
+func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
+	var certData []byte
+	if v, ok := secret.Data["tls.crt"]; !ok {
+		return nil
+	} else {
+		certData = v
+	}
+
+	for len(certData) > 0 {
+		block, data := pem.Decode(certData)
+		if block == nil {
+			break
+		}
+		certData = data
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			continue
+		}
+		foundSAN := false
+		for i := range cert.DNSNames {
+			if cert.DNSNames[i] == domain {
+				foundSAN = true
+			}
+		}
+		if cert.Subject.CommonName == domain && !foundSAN {
```

Comment:
Is there always going to be a common name, even if there is a SAN?

Reply:
No, but what we really care about is the situation where a serving certificate is accepted by older clients and rejected by newer clients.  Checking that the CN matches the ingresscontroller domain and that no SAN does is the best approach I've come up with to warn when there is a problem without causing false positives.  Consider these four cases:

* The certificate has no SAN and no CN; such a certificate wouldn't work at all as a serving certificate even for older clients, so we should ignore it.
* The certificate has no CN and does have a SAN; this is fine: if the SAN is valid, the certificate will be accepted by both old clients as well as new clients (and if the SAN is invalid, it won't be accepted by old or new clients), so we should ignore it.
* The certificate has a CN that doesn't match the ingresscontroller domain; in this case, the certificate is presumably not the serving certificate but rather an intermediate, and anyway, as with the previous case, either it has a valid SAN, in which case both old as well as new clients will accept it, or it does not, in which case both old as well as new clients will reject it, so we should ignore it.
* The certificate has a CN that does match the ingresscontroller domain; in this case, the certificate most likely is the serving certificate, and we need to make sure it has a SAN with the same domain in order to still be accepted by newer Go clients.  

The last case is really the only one that matters for upgrades, so it is the only one where we set `Upgradeable=True` if we find a mismatch.